### PR TITLE
Add quantity

### LINKED ISSUE #81: As a user, I can add the weight or quantity of each food item I add into the fridge
so that I know the amount of the specific food item I have left

### LINKED ISSUE #82: Edit storage to account for weight/quantity

### LINKED ISSUE #86: Check other classes for any edits regarding weight/quantity

### PR DESCRIPTION
**Pull request by SimJJ96**
Fixes #81 
Fixes #82
Fixes #86

### CODE REVIEW

File: src/main/java/seedu/fridgefriend/food/Weight.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+public class Weight extends Quantity {
+
+    private String unit;
+    public static final String UNIT_CHARACTER = "g";
+
+    public Weight(String quantityString) throws InvalidQuantityException {
+        try {
+            String[] words = quantityString.split("(?<=\\d)(?=[a-z])");
+            this.number = Integer.parseInt(words[0]);
```
Comment (Leeyp) [inline src/main/java/seedu/fridgefriend/food/Weight.java]: Might be good to clarify what this split command does in a javadoc comment or similar.

File: src/main/java/seedu/fridgefriend/food/Weight.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            this.unit = words[1];
+            if (unit.length() > 1 || !unit.equals("g")) {
+                throw new InvalidQuantityException();
```
Comment (Leeyp) [inline src/main/java/seedu/fridgefriend/food/Weight.java]: Would be good to extract this out as a method for Code Quality purposes

**Review by Leeyp — CHANGES_REQUESTED** [review]

**Review by Leeyp — APPROVED** [review]
LGTM

File: src/main/java/seedu/fridgefriend/command/AddCommand.java [hunk 3]
```diff
@@ -20,22 +22,24 @@
...
+ * and quantity are necessary fields
  */
 public class AddCommand extends Command {
 
     private final Food foodToAdd;
 
     public AddCommand(String foodName, FoodCategory category, String expiryString,
-                      FoodStorageLocation location) throws InvalidDateException {
+                      FoodStorageLocation location, Quantity quantity)
+            throws InvalidDateException, InvalidQuantityException {
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/command/AddCommand.java]: This line should have extra spacing

File: src/main/java/seedu/fridgefriend/command/AddCommand.java [hunk 4]
```diff
@@ -68,52 +72,53 @@ public String getMessagePrintedToUser() {
     }
 
     public static Food categoriseAndGenerateFood(String foodName, FoodCategory category,
-                                                 String expiryString, FoodStorageLocation location)
-            throws InvalidDateException {
+                                                 String expiryString, FoodStorageLocation location,
+                                                 Quantity quantity)
+            throws InvalidDateException, InvalidQuantityException {
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/command/AddCommand.java]: lines 75-77 has weird spacing

File: src/main/java/seedu/fridgefriend/food/Food.java [hunk 5]
```diff
@@ -1,37 +1,47 @@
...
     protected FoodCategory category;
     protected String foodName;
     protected ExpiryDate expiryDate;
     protected FoodStorageLocation storageLocation;
+    protected Quantity quantity;
 
     public Food(String foodName, FoodCategory category, String expiryString,
-            FoodStorageLocation storageLocation) throws InvalidDateException {
+                FoodStorageLocation storageLocation, Quantity quantity)
+            throws InvalidDateException, InvalidQuantityException {
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/food/Food.java]: This line should have extra spacing

File: src/main/java/seedu/fridgefriend/food/categories/CookedDish.java [hunk 6]
```diff
@@ -1,14 +1,17 @@
...
 import seedu.fridgefriend.food.FoodStorageLocation;
+import seedu.fridgefriend.food.Quantity;
 
 public class CookedDish extends Food {
 
     public CookedDish(String foodName, FoodCategory category, String expiryString,
-                      FoodStorageLocation storageLocation) throws InvalidDateException {
-        super(foodName, category, expiryString, storageLocation);
+                      FoodStorageLocation storageLocation, Quantity quantity) throws InvalidDateException,
+            InvalidQuantityException {
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/food/categories/CookedDish.java]: this line should have extra space

File: src/main/java/seedu/fridgefriend/food/categories/Dairy.java [hunk 7]
```diff
@@ -1,14 +1,17 @@
...
 import seedu.fridgefriend.food.FoodStorageLocation;
+import seedu.fridgefriend.food.Quantity;
 
 public class Dairy extends Food {
 
     public Dairy(String foodName, FoodCategory category, String expiryString,
-                 FoodStorageLocation storageLocation) throws InvalidDateException {
-        super(foodName, category, expiryString, storageLocation);
+                 FoodStorageLocation storageLocation, Quantity quantity) throws InvalidDateException,
+            InvalidQuantityException {
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/food/categories/Dairy.java]: this line should have extra spaces

File: src/main/java/seedu/fridgefriend/food/categories/Egg.java [hunk 8]
```diff
@@ -1,14 +1,17 @@
...
 import seedu.fridgefriend.food.FoodStorageLocation;
+import seedu.fridgefriend.food.Quantity;
 
 public class Egg extends Food {
 
     public Egg(String foodName, FoodCategory category, String expiryString,
-               FoodStorageLocation storageLocation) throws InvalidDateException {
-        super(foodName, category, expiryString, storageLocation);
+               FoodStorageLocation storageLocation, Quantity quantity) throws InvalidDateException,
+            InvalidQuantityException {
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/food/categories/Egg.java]: this line should have extra spaces

File: src/main/java/seedu/fridgefriend/food/categories/Frozen.java [hunk 9]
```diff
@@ -1,14 +1,17 @@
...
 import seedu.fridgefriend.food.FoodStorageLocation;
+import seedu.fridgefriend.food.Quantity;
 
 public class Frozen extends Food {
 
     public Frozen(String foodName, FoodCategory category, String expiryString,
-                  FoodStorageLocation storageLocation) throws InvalidDateException {
-        super(foodName, category, expiryString, storageLocation);
+                  FoodStorageLocation storageLocation, Quantity quantity) throws InvalidDateException,
+            InvalidQuantityException {
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/food/categories/Frozen.java]: this line should have extra spaces

File: src/main/java/seedu/fridgefriend/food/categories/Fruit.java [hunk 10]
```diff
@@ -1,14 +1,17 @@
...
 import seedu.fridgefriend.food.FoodStorageLocation;
+import seedu.fridgefriend.food.Quantity;
 
 public class Fruit extends Food {
 
     public Fruit(String foodName, FoodCategory category, String expiryString,
-                 FoodStorageLocation storageLocation) throws InvalidDateException {
-        super(foodName, category, expiryString, storageLocation);
+                 FoodStorageLocation storageLocation, Quantity quantity) throws InvalidDateException,
+            InvalidQuantityException {
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/food/categories/Fruit.java]: this line should have extra spaces

File: src/main/java/seedu/fridgefriend/food/categories/Meat.java [hunk 11]
```diff
@@ -1,14 +1,17 @@
...
 import seedu.fridgefriend.food.FoodStorageLocation;
+import seedu.fridgefriend.food.Quantity;
 
 public class Meat extends Food {
 
     public Meat(String foodName, FoodCategory category, String expiryString,
-                FoodStorageLocation storageLocation) throws InvalidDateException {
-        super(foodName, category, expiryString, storageLocation);
+                FoodStorageLocation storageLocation, Quantity quantity) throws InvalidDateException,
+            InvalidQuantityException {
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/food/categories/Meat.java]: this line should have extra spaces

File: src/main/java/seedu/fridgefriend/food/categories/Other.java [hunk 12]
```diff
@@ -1,14 +1,17 @@
...
 import seedu.fridgefriend.food.FoodStorageLocation;
+import seedu.fridgefriend.food.Quantity;
 
 public class Other extends Food {
 
     public Other(String foodName, FoodCategory category, String expiryString,
-                 FoodStorageLocation storageLocation) throws InvalidDateException {
-        super(foodName, category, expiryString, storageLocation);
+                 FoodStorageLocation storageLocation, Quantity quantity) throws InvalidDateException,
+            InvalidQuantityException {
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/food/categories/Other.java]: this line should have extra spaces

File: src/main/java/seedu/fridgefriend/food/categories/ReadyToEat.java [hunk 13]
```diff
@@ -1,14 +1,17 @@
...
 import seedu.fridgefriend.food.FoodStorageLocation;
+import seedu.fridgefriend.food.Quantity;
 
 public class ReadyToEat extends Food {
 
     public ReadyToEat(String foodName, FoodCategory category, String expiryString,
-                      FoodStorageLocation storageLocation) throws InvalidDateException {
-        super(foodName, category, expiryString, storageLocation);
+                      FoodStorageLocation storageLocation, Quantity quantity) throws InvalidDateException,
+            InvalidQuantityException {
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/food/categories/ReadyToEat.java]: this line should have extra spaces

File: src/main/java/seedu/fridgefriend/food/categories/Seafood.java [hunk 14]
```diff
@@ -1,14 +1,17 @@
...
 import seedu.fridgefriend.food.FoodStorageLocation;
+import seedu.fridgefriend.food.Quantity;
 
 public class Seafood extends Food {
 
     public Seafood(String foodName, FoodCategory category, String expiryString,
-                   FoodStorageLocation storageLocation) throws InvalidDateException {
-        super(foodName, category, expiryString, storageLocation);
+                   FoodStorageLocation storageLocation, Quantity quantity) throws InvalidDateException,
+            InvalidQuantityException {
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/food/categories/Seafood.java]: this line should have extra spaces

File: src/main/java/seedu/fridgefriend/food/categories/Vegetable.java [hunk 15]
```diff
@@ -1,14 +1,17 @@
...
+import seedu.fridgefriend.food.Quantity;
 
 public class Vegetable extends Food {
 
     public Vegetable(String foodName, FoodCategory category, String expiryString,
-                     FoodStorageLocation storageLocation) throws InvalidDateException {
-        super(foodName, category, expiryString, storageLocation);
+                     FoodStorageLocation storageLocation, Quantity quantity) throws InvalidDateException,
+            InvalidQuantityException {
+        super(foodName, category, expiryString, storageLocation, quantity);
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/food/categories/Vegetable.java:13]: this line should have extra spaces

File: src/main/java/seedu/fridgefriend/utilities/Parser.java [hunk 16]
```diff
@@ -204,23 +238,25 @@ public static Command getSearchCommand(String description) throws EmptyDescripti
...
-
     private static Command getExpiringCommand() {
         Command expiringCommand = new ExpiringCommand();
         return expiringCommand;
     }
 
     /**
      * Returns a HelpCommand object.
+     *
+     * @return help instructions
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/utilities/Parser.java]: technically its a help command object

File: src/main/java/seedu/fridgefriend/utilities/Parser.java [hunk 17]
```diff
@@ -204,23 +238,25 @@ public static Command getSearchCommand(String description) throws EmptyDescripti
...
-
     public static Command getHelpCommand() {
         Command helpCommand = new HelpCommand();
         return helpCommand;
     }
 
     /**
      * Returns a ByeCommand object.
+     *
+     * @return acknowledgement to exit the program
```
Comment (kwokyto) [inline src/main/java/seedu/fridgefriend/utilities/Parser.java]: technically its an bye command object

**Review by kwokyto — APPROVED** [review]
Other than the cosmetic comments, LGTM

File: src/main/java/seedu/fridgefriend/utilities/Parser.java [hunk 18]
```diff
@@ -113,15 +118,28 @@ public static Command parseCommand(String[] parsedInput)
...
-                    + " /loc (?<storageLocation>[^/]+)");
+                    + " /loc (?<storageLocation>[^/]+)"
+                    + " /qty (?<quantity>[^/]+)");
+
+    /**
+     * Define arguments format for add food command with weight.
+     * A Pattern object which defines how the input string for food item
+     * that should look like. [^/]+ implies 1 or more characters except for '/'
+     */
+    public static final Pattern FOOD_DATA_ARGS_FORMAT_WEIGHT =
```
Comment (Vinci-Hu) [inline src/main/java/seedu/fridgefriend/utilities/Parser.java]: Do you plan to add a pattern for every child class of QUANTITY? But current implementation can work.

**Review by Vinci-Hu — APPROVED** [review]
Does not seem very OOP but work-able. If we need to change the structure it may require a discussion among group. I will try to continue working on the current structure.

File: src/main/java/seedu/fridgefriend/utilities/Parser.java (hunk 18 above)
Comment (SimJJ96) [reply src/main/java/seedu/fridgefriend/utilities/Parser.java]: No. I was thinking that these two would be sufficient as we would have the user enter the amount either by /qty or /wgt only

**Pull request merged**